connectors/sources/microsoft_teams.py

```diff
@@ -0,0 +1,340 @@
+            self._logger.debug(f"Calling Microsoft Teams Endpoint: {absolute_url}")
+            async with self._http_session.get(
+                url=absolute_url,
+                headers={"authorization": f"Bearer {token}"},
+            ) as resp:
+                yield resp
+        except aiohttp.client_exceptions.ClientOSError:
+            self._logger.error(
+                "Graph API dropped the connection. It might indicate, that connector makes too many requests - decrease concurrency settings, otherwise Graph API can block this app."
+            )
+            raise
+        except ClientResponseError as e:
+            await self._handle_client_response_error(absolute_url, e)
+
+    async def _handle_client_response_error(self, absolute_url, e):
+        if e.status == 403:
+            raise PermissionsMissing(
+                f"Received Unauthorized response for {absolute_url}.\nVerify that the correct Graph API and Microsoft Teams permissions are granted to the app and admin consent is given. If the permissions and consent are correct, wait for several minutes and try again."
+            ) from e
+        elif e.status == 404:
+            raise NotFound from e
+        elif e.status == 500:
+            raise InternalServerError(
+                f"Received InternalServerError error for {absolute_url}: Exception: {e}"
+            ) from e
+        else:
+            raise
+
+    async def ping(self):
+        return await self.fetch(
+            url=URLS[UserEndpointName.PING.value].format(base_url=BASE_URL)
+        )
+
+    async def close(self):
+        self._sleeps.cancel()
+        await self._http_session.close()
+
+
+class MicrosoftTeamsDataSource(BaseDataSource):
+    """Microsoft Teams"""
+
+    name = "Microsoft Teams"
+    service_type = "microsoft_teams"
+
+    def __init__(self, configuration):
+        """Set up the connection to the Microsoft Teams.
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+
+    def _set_internal_logger(self):
+        self.client.set_logger(self._logger)
+
+    @cached_property
+    def client(self):
+        tenant_id = self.configuration["tenant_id"]
+        client_id = self.configuration["client_id"]
+        client_secret = self.configuration["secret_value"]
+        username = self.configuration["username"]
+        password = self.configuration["password"]
+
+        return MicrosoftTeamsClient(
+            tenant_id, client_id, client_secret, username, password
+        )
+
+    async def validate_config(self):
+        await super().validate_config()
+
+        # Check that we can log in into Graph API
+        await self.client._api_token.get_with_username_password()
+
+    async def ping(self):
+        """Verify the connection with Microsoft Teams"""
+        try:
+            await self.client.ping()
```

Comment:
I'm not sure whether this works. The call flow is `ping` -> `fetch` -> `_get_json` -> `_get`. `_get_json` is a generator, and `fetch` and `ping` just `await` and `return`. Do you need to add an `anext` to make it execute?

Reply:
This is working fine as expected. But We can add anext to _get_json as it returns the single response only.